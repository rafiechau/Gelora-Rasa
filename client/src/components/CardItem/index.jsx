import { Box, Button, Card, CardActionArea, CardActions, CardContent, CardMedia, Typography } from '@mui/material';
import PropTypes from 'prop-types';
import ShareIcon from '@mui/icons-material/Share';
import toast from 'react-hot-toast';
import config from '@config/index';
import { useNavigate } from 'react-router-dom';

const CardItem = ({ event }) => {
  const navigate = useNavigate();

  const navigateToDetailEvent = () => {
    navigate(`/detail/${event.id}`);
  };

  const formattedPrice = new Intl.NumberFormat('id-ID', {
    style: 'currency',
    currency: 'IDR',
    minimumFractionDigits: 0,
  }).format(event?.price);

  const formattedDate = new Intl.DateTimeFormat('id-ID', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  }).format(new Date(event?.date));

  const imageUrl =
    event && event?.image
      ? `${config.api.server}${event?.image.replace('\\', '/')}`
      : 'https://source.unsplash.com/random';

  const handleShareClick = async () => {
    const url = `${window.location.origin}/detail/${event.id}`;
    const text = `Check out this event: ${event.eventName}`; // Teks yang ingin dibagikan

    try {
      // Cek apakah Web Share API tersedia
      if (navigator.share) {
        await navigator.share({
          title: event.eventName,
          text,
          url,
        });
        toast.success('Event shared successfully!');
      } else {
        // Jika Web Share API tidak tersedia, gunakan clipboard
        await navigator.clipboard.writeText(url);
        toast.success('Link copied to clipboard!');
      }
    } catch (err) {
      toast.error('Failed to share event.');
    }
  };

  return (
    <Card
      data-testid="CardItem"
      sx={{
        maxWidth: 345,
        width: '100%',
        margin: 2,
        boxShadow: '0 8px 16px 0 rgba(0,0,0,0.2)',
        transition: '0.3s',
        '&:hover': {
          boxShadow: '0 16px 32px 0 rgba(0,0,0,0.2)',
        },
        borderRadius: '8px',
        overflow: 'hidden',
      }}
    >
      <CardActionArea data-testid="card-action-area" onClick={() => navigateToDetailEvent(event.id)}>
        <CardMedia
          data-testid="card-media"
          component="img"
          height="200"
          image={imageUrl}
          alt={event.eventName}
          sx={{ objectFit: 'cover', width: '100%' }}
        />
        <CardContent data-testid="card-content" sx={{ bgcolor: 'rgba(29, 36, 51, 0.9)', color: 'white' }}>
          <Typography gutterBottom variant="h6" component="div" sx={{ fontWeight: 'bold' }}>
            {event.eventName}
          </Typography>
          <Box display="flex" justifyContent="space-between" alignItems="center" mt={1} gap={2}>
            <Box>
              <Typography variant="caption" display="block" color="text.secondary">
                {formattedDate}
              </Typography>
              <Typography variant="caption" display="block" color="text.secondary">
                {event?.Location?.namaProvinsi}
              </Typography>
            </Box>
            <Typography
              variant="subtitle1"
              sx={{ display: 'flex', alignItems: 'center', color: '#4DD0E1', fontWeight: 'medium' }}
            >
              {formattedPrice}
            </Typography>
          </Box>
        </CardContent>
      </CardActionArea>
      <CardActions
        sx={{
          bgcolor: '#16202A',
          padding: '16px',
          justifyContent: 'space-between',
        }}
      >
        <Button size="small" sx={{ color: 'white' }} startIcon={<ShareIcon />} onClick={handleShareClick}>
          Share
        </Button>
      </CardActions>
    </Card>
  );
};

CardItem.propTypes = {
  event: PropTypes.object.isRequired,
};

export default CardItem;
