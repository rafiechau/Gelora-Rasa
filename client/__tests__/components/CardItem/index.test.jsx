import { render, fireEvent, screen } from '@testing-library/react';
import CardItem from '@components/CardItem';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

const mockEvent = {
  id: 1,
  eventName: 'Sample Event',
  date: '2023-01-01',
  price: 10000,
  image: '/path/to/image.jpg',
  Location: {
    namaProvinsi: 'Some Province',
  },
};

let wrapper;

beforeEach(() => {
  wrapper = render(<CardItem event={mockEvent} />);
});

describe('CardItem', () => {
  test('Should render correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('CardItem')).toBeInTheDocument();
  });

  test('Should navigate to detail page on card click', () => {
    const { getByTestId } = wrapper;
    const cardActionArea = getByTestId('card-action-area');
    fireEvent.click(cardActionArea);
    expect(mockNavigate).toHaveBeenCalledWith(`/detail/${mockEvent.id}`);
  });

  test('Should render event name correctly', () => {
    expect(screen.getByText(mockEvent.eventName)).toBeInTheDocument();
  });

  test('Should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
