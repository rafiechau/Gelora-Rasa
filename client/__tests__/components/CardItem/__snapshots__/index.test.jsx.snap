// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1cnxcpd-MuiPaper-root-MuiCard-root"
        data-testid="CardItem"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
          data-testid="card-action-area"
          tabindex="0"
          type="button"
        >
          <img
            alt="Sample Event"
            class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-xx7l7s-MuiCardMedia-root"
            data-testid="card-media"
            height="300"
            src="http://localhost:5000//path/to/image.jpg"
          />
          <div
            class="MuiCardContent-root css-uhve52-MuiCardContent-root"
            data-testid="card-content"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-sskem8-MuiTypography-root"
            >
              Sample Event
            </div>
            <div
              class="MuiBox-root css-ldta6s"
            >
              <div
                class="MuiBox-root css-0"
              >
                <span
                  class="MuiTypography-root MuiTypography-caption css-1viqp7r-MuiTypography-root"
                >
                  1 Januari 2023
                </span>
                <span
                  class="MuiTypography-root MuiTypography-caption css-1viqp7r-MuiTypography-root"
                >
                  Some Province
                </span>
              </div>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-77lqui-MuiTypography-root"
              >
                Rp 10.000
              </h6>
            </div>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1bo3tks-MuiCardActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-uj3a13-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ShareIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
                />
              </svg>
            </span>
            app_btn_share
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1cnxcpd-MuiPaper-root-MuiCard-root"
      data-testid="CardItem"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
        data-testid="card-action-area"
        tabindex="0"
        type="button"
      >
        <img
          alt="Sample Event"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-xx7l7s-MuiCardMedia-root"
          data-testid="card-media"
          height="300"
          src="http://localhost:5000//path/to/image.jpg"
        />
        <div
          class="MuiCardContent-root css-uhve52-MuiCardContent-root"
          data-testid="card-content"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-sskem8-MuiTypography-root"
          >
            Sample Event
          </div>
          <div
            class="MuiBox-root css-ldta6s"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiTypography-root MuiTypography-caption css-1viqp7r-MuiTypography-root"
              >
                1 Januari 2023
              </span>
              <span
                class="MuiTypography-root MuiTypography-caption css-1viqp7r-MuiTypography-root"
              >
                Some Province
              </span>
            </div>
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-77lqui-MuiTypography-root"
            >
              Rp 10.000
            </h6>
          </div>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1bo3tks-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-uj3a13-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ShareIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
              />
            </svg>
          </span>
          app_btn_share
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
